f3/granite.go

```diff
@@ -424,22 +440,23 @@ func (i *instance) tryPrepare() {
 func (i *instance) beginCommit() {
 	i.phase = COMMIT
 	i.phaseTimeout = i.alarmAfterSynchrony(PREPARE)
-	i.broadcast(COMMIT, i.value, nil)
+	i.broadcast(COMMIT, i.value, nil, i.round)
 }
 
 func (i *instance) tryCommit(round uint32) {
 	// Unlike all other phases, the COMMIT phase stays open to new messages even after an initial quorum is reached,
 	// and the algorithm moves on to the next round.
 	// A subsequent COMMIT message can cause the node to decide, so there is no check on the current phase.
 	committed := i.roundState(round).committed
-	foundQuorum := committed.ListQuorumAgreedValues()
+	foundQuorum := committed.ListStrongQuorumAgreedValues()
 	timeoutExpired := i.host.Time() >= i.phaseTimeout
 
-	if len(foundQuorum) > 0 && !foundQuorum[0].IsZero() {
+	if len(foundQuorum) > 0 && !foundQuorum[0].IsZero() && !i.decideSent {
```

Comment:
This `!decideSent` is not in the algorithm [here](https://hackmd.io/YDW-1jd7Tzy8pp2kb6T7zQ). It's not obvious what to do here - the Bracha amplification introduces a problem. Intuitively a strong quorum of COMMIT is better than a weak quorum of DECIDE, but here it is losing out (obviously with some adversarial equivocation)!

Leaving this condition here will prevent the phase transition resulting from the strong quorum of commits. For now I think it would be better to remove the condition, but perhaps file an issue for it. (One alternative to note in the issue: we could push the condition down into the `beginDecide` method, allowing the state transition but not the broadcast. This would put a node in a weird state where it can't contribute to the strong quorum of DECIDE for the value it COMMITTED, and is waiting for a strong quorum of DECIDE for a different value).

I hope we can just remove the amplification.

Reply:
A strong quorum of COMMIT of any round (and noting that all honest nodes send a COMMIT before they move on to the next round) is actually as good as a weak quorum of DECIDE without rounds (and noting that honest nodes *only* send DECIDE when they have received a strong quorum of COMMITs). 

Strictly speaking they are just as good: an adversary controlling less than a weak quorum cannot fake either and an adversary controlling a weak quorum or more can fake both. 

However, even without Bracha, I think `!i.decideSent` makes sense to prevent nodes from broadcasting the same DECIDE twice (even without bracha they may send it after receiving a strong quorum of DECIDE, which may be delivered before a strong quorum of COMMIT). 

The modification of the algorithm that includes i.decideSent (and the Bracha amplification) is the one in the [PR of the FIP](https://github.com/filecoin-project/FIPs/blob/9ad97ce63e4d37b60ed9077ecc0941a08b521e39/FIPS/fip-xxxx.md?plain=1#L407). AFAIK this is the most up to date version of the pseudocode and the only one that is currently being updated with new comments.

Reply:
Actually, since we are repurposing `i.phase = DECIDE` and introducing `i.phase = TERMINATED`, we can instead use `i.phase != DECIDE && i.phase != TERMINATED` and completely forget about `i.decideSent`. (Done in the upcoming pushed changes).

Reply:
And actually `i.phase != TERMINATED` is only necessary for clarity, as tryCommit will only happen if `i.phase != TERMINATED` in the first place with the current code. Happy to remove it but leaving it for clarity for now.

Reply:
Right, i.decideSent is redundant because the phase transitions ensure each messages is only sent once. Defensive programming of extra, redundant checks will confuse things.